lib/bolt/result_set.rb

```diff
@@ -0,0 +1,94 @@
+module Bolt
+  class ResultSet
+    attr_reader :results
+
+    include Enumerable
+
+    # We only want want to include these when puppet is loaded
+    def self.include_iterable
+      include(Puppet::Pops::Types::Iterable)
+      include(Puppet::Pops::Types::IteratorProducer)
+    end
+
+    def iterator
+      if Object.const_defined?(:Puppet) && Puppet.const_defined?(:Pops) &&
+         self.class.included_modules.include?(Puppet::Pops::Types::Iterable)
+        Puppet::Pops::Types::Iterable.on(@results, Bolt::Result)
+      else
+        raise NotImplementedError, "iterator requires puppet code to be loaded."
+      end
+    end
+
+    def initialize(results)
+      @results = results
+    end
+
+    def each
+      @results.each { |r| yield r }
+    end
+
+    def result_hash
+      @result_hash ||= @results.each_with_object({}) do |result, acc|
+        acc[result.target.name] = result
+      end
+    end
+
+    def count
+      @results.size
+    end
+    alias length count
+    alias size count
+
+    def empty
+      @results.empty?
+    end
+    alias empty? empty
+
+    def targets
+      results.map(&:target)
+    end
+
+    def names
+      @results.map { |r| r.target.name }
+    end
+
+    def ok
+      @results.all?(&:ok?)
+    end
+    alias ok? ok
+
+    def error_set
+      filtered = @results.reject(&:ok?)
+      ResultSet.new(filtered)
+    end
+
+    def ok_set
+      filtered = @results.select(&:success?)
+      self.class.new(filtered)
+    end
+
+    def find(target_name)
+      result_hash[target_name]
+    end
+
+    def first
+      @results.first
+    end
+
+    def eql?(other)
+      self.class == other.class && @results == other.results
+    end
+
+    def to_json
```

Comment:
Getting an error
```
Plan aborted: run_task 'upgrade_pe::monolithic' failed on 1 nodes
bundler: failed to load command: bolt (/Users/michaelsmith/puppetlabs/orchestration-control/.bundle/gems/ruby/2.3.0/bin/bolt)
ArgumentError: wrong number of arguments (given 1, expected 0)
  /Users/michaelsmith/puppetlabs/orchestration-control/.bundle/gems/ruby/2.3.0/bundler/gems/bolt-f260b52f88e9/lib/bolt/result_set.rb:82:in `to_json'
  /Users/michaelsmith/.rbenv/versions/2.3.4/lib/ruby/2.3.0/json/common.rb:286:in `generate'
  /Users/michaelsmith/.rbenv/versions/2.3.4/lib/ruby/2.3.0/json/common.rb:286:in `pretty_generate'
  /Users/michaelsmith/puppetlabs/orchestration-control/.bundle/gems/ruby/2.3.0/bundler/gems/bolt-f260b52f88e9/lib/bolt/outputter/human.rb:142:in `fatal_error'
```

Plan I'm trying to run is
```
plan upgrade_pe::mono_cms(
  String[1]                                      $mono_master,
  String[1]                                      $build,
  Optional[Variant[String[1], Array[String[1]]]] $compile_masters = undef,
) {
  $mono = run_task('upgrade_pe::monolithic', $mono_master, build => $build)
  $mono.each |$result| {
    if $result.ok {
      notice("${result.target} completed:\n${result.message}")
    } else {
      notice("${result.target} errored with message ${result.error}")
    }
  }

  if ($mono.ok and $compile_masters) {
    $_compile_masters = $compile_masters.split(',')
    $cms = run_task('upgrade_pe::compile_master', $_compile_masters, master => $mono_master)
    $cms.each |$result| {
      if $result.ok {
        notice("${result.target} completed:\n${result.message}")
      } else {
        notice("${result.target} errored with message ${result.error}")
      }
    }
  }
}
```

Reply:
this was caused by to_json not accepting opts. it should be fixed now